### PR TITLE
Replace Global minimum_mysql_version with adapter-specific checks

### PR DESCRIPTION
**Pull request by chippison**
### Description
Dev-19321

Please include a description of this change and which issue it fixes. If no issue exists yet please include context and what problem it solves.

### Checklist

- [NA] I have understood, reviewed, and tested all AI outputs before use
- [NA] All AI instructions respect security, IP, and privacy rules

### Review

- [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
- [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behaviour of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
- [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
- [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
- [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
- [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
- [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
- [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
- [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
- [ ] [New documentation added or existing documentation updated](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)

### CODE REVIEW

File: config/global.ini.php [hunk 1]
```diff
@@ -427,11 +427,6 @@
 ; you can prevent this action from happening by setting this parameter to value bigger than 1
 purge_date_range_archives_after_X_days = 1
 
-; MySQL minimum required version
-; note: timezone support added in 4.1.3
-minimum_mysql_version = 4.1
-
-
```
Comment (chippison) [inline config/global.ini.php:434]: Removed as this is no longer used anywhere else in the code

File: plugins/Installation/Controller.php [hunk 2]
```diff
@@ -157,14 +157,13 @@ public function databaseSetup()
         if ($form->validate()) {
             try {
                 $dbInfos = $form->createDatabaseObject();
+                $this->createConfigFile($dbInfos);
```
Comment (chippison) [inline plugins/Installation/Controller.php:160]: We needed to create the config file first before checking so that we can capture the latest selection from the user. When this was after the checkDatabaseVersion() function, it was not getting the values if you changed your initial selection after the first exception (first selection had higher minimum version than required)

File: plugins/Installation/Controller.php [hunk 3]
```diff
@@ -157,14 +157,13 @@ public function databaseSetup()
...
 
+                DbHelper::unsetSchema();
```
Comment (chippison) [inline plugins/Installation/Controller.php]: We need to unset the initial schema if there was one so that when we compare versions, it is the one selected ($dbInfos) instead of the previous schema if there was one

File: core/Db/Schema/Mysql.php [hunk 4]
```diff
@@ -30,13 +30,17 @@ class Mysql implements SchemaInterface
...
     public const MAX_TABLE_NAME_LENGTH = 64;
-
     private $tablesInstalled = null;
+    protected $minimumSupportedVersion = '5.5';
 
     public function getDatabaseType(): string
     {
         return 'MySQL';
     }
+    public function getMinimumSupportedVersion(): string
```
Comment (michalkleiner) [inline core/Db/Schema/Mysql.php:41]:
```suggestion
    
    public function getMinimumSupportedVersion(): string
```

File: core/Db/Schema/Mariadb.php [hunk 5]
```diff
@@ -16,6 +16,7 @@
  */
 class Mariadb extends Mysql
 {
+    protected $minimumSupportedVersion = '5.5';
```
Comment (michalkleiner) [inline core/Db/Schema/Mariadb.php:19]:
```suggestion
    protected $minimumSupportedVersion = '5.5';
    
```

File: tests/PHPUnit/Unit/Db/Adapter/Pdo/MysqlTest.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        $adapter->checkServerVersion();
+        $this->addToAssertionCount(1);
+    }
+
+    /**
+     * This will 'mock' the Schema class to return a specific minimum version.
+     * @param string $minimumVersion
+     * @return Schema
+     */
+    private function createSchemaDouble(string $minimumVersion): Schema
```
Comment (michalkleiner) [inline tests/PHPUnit/Unit/Db/Adapter/Pdo/MysqlTest.php]: If it's creating a mock, we can name the method (and the one below similarly) `createMockSchema` and `createMockAdapter`

File: core/DbHelper.php [hunk 7]
```diff
@@ -465,4 +465,13 @@ public static function isValidDbname($dbname)
     {
         return (0 !== preg_match('/(^[a-zA-Z0-9]+([a-zA-Z0-9\_\.\-\+]*))$/D', $dbname));
     }
+
+    /**
+     * Helper funtion to unset the schema
+     * @return void
+     */
+    public static function unsetSchema()
```
Comment (sgiehl) [inline core/DbHelper.php]:
```suggestion
    public static function unsetSchema(): void
```

File: tests/PHPUnit/Unit/Db/Adapter/Pdo/MysqlTest.php (hunk 6 above)
Comment (chippison) [reply tests/PHPUnit/Unit/Db/Adapter/Pdo/MysqlTest.php]: This is all good now

File: plugins/Installation/Controller.php (hunk 3 above)
Comment (caddoo) [reply plugins/Installation/Controller.php]: Optional feedback: If you always need to unset the Schema when checking the database version, why not put it in the checkDatabaseVersion function call and remove the additional public unsetSchema method.

File: tests/PHPUnit/Unit/Db/Adapter/MysqliTest.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+namespace Piwik\Tests\Unit\Db\Adapter;
+
+use Exception;
+use PHPUnit\Framework\TestCase;
+use Piwik\Db\Adapter\Mysqli;
+use Piwik\Db\Schema;
+
+class MysqliTest extends TestCase
```
Comment (caddoo) [inline tests/PHPUnit/Unit/Db/Adapter/MysqliTest.php:17]: Nice clean tests 👍

File: plugins/Installation/Controller.php (hunk 2 above)
Comment (caddoo) [reply plugins/Installation/Controller.php:160]: I couldn't replicate this problem, what initial selection did it not update?

Can jump in call if needed to show me.

image

Comment (chippison) [reply plugins/Installation/Controller.php:160]: This happens if you have two different 'minimum_supported_version' for Mysql and MariaDB
 i.e
 - Mysql minimum version set is 5.8
 - MariaDB minimum version set is 5.6
 - if you have db installed that is less lets say 5.5
How to reproduce:
 - first try to choose DB engine to be MySql, it would error out saying minimum version is 5.8
 - if you then change/select DB engine to be MariaDb, before this change, it would still error out saying the minimum version needed is 5.8 (Mysql minimum version instead of MariaDb).
 - this happens because it still takes the initial choice because the new selection was not saved.

This is a little bit hard to replicate as this needs the dev to change 'minimum_supported_version' locally to get test this edge case. Can talk about more on this in a call if needed

**Review by caddoo — DISMISSED** [review]

**Review by sgiehl — DISMISSED** [review]
Tested this locally and it seems to work as expected. Code overall also looks good 👍

**Review by sgiehl — APPROVED** [review]
